### PR TITLE
[Downgrade PHP 8.0]  Add DowngradeTrailingCommasInParamUseRector

### LINKED ISSUE #4160: [Downgrade PHP 8.0] Trailing comma in parameter lists
# Feature Request

Downgrade new feature from PHP 8.0 to its equivalent PHP 7.4 code
 
## Diff 

```diff
public function(
    string $parameterA,
    int $parameterB,
-    Foo $objectfoo,
+    Foo $objectfoo
) {
    // …
}
```

### PR DESCRIPTION
**Pull request by samsonasik**
Fixes #4160

### CODE REVIEW

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba it is ready for review.

File: rules/downgrade-php80/src/Rector/ClassMethod/DowngradeTrailingCommasInParamUseRector.php [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    {
+        return [ClassMethod::class, Function_::class, Closure::class];
+    }
+
+    /**
+     * @param ClassMethod|Function_|Closure $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        if (! $node instanceof Closure) {
```
Comment (simivar) [inline rules/downgrade-php80/src/Rector/ClassMethod/DowngradeTrailingCommasInParamUseRector.php:78]: We could skip the negation here and run `$this->processUses` only when the instance is `Closure`. Thanks to that we will only have one `return` statement and a simpler condition (without negation).

File: rules/downgrade-php80/src/Rector/ClassMethod/DowngradeTrailingCommasInParamUseRector.php [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+            return $this->processParams($node);
+        }
+
+        $node = $this->processUses($node);
+        return $this->processParams($node);
+    }
+
+    private function processParams(Node $node): ?Node
+    {
+        if ($node->params !== []) {
```
Comment (simivar) [inline rules/downgrade-php80/src/Rector/ClassMethod/DowngradeTrailingCommasInParamUseRector.php:88]: Here as well we could flip the condition and return immediately `null` when `$node->params === []` and do the rest outside of if statement.

File: rules/downgrade-php80/src/Rector/ClassMethod/DowngradeTrailingCommasInParamUseRector.php (hunk 1 above)
Comment (TomasVotruba) [reply rules/downgrade-php80/src/Rector/ClassMethod/DowngradeTrailingCommasInParamUseRector.php:78]: I'm not sure what exactly you mean, the code in words is very hard to read.
Could you maybe diff a suggestion?

Like:

```suggestion
-        if (! $node instanceof Closure) {
+        if (! $node instanceof Closure....) {
```

File: rules/downgrade-php80/src/Rector/ClassMethod/DowngradeTrailingCommasInParamUseRector.php [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+     * @param ClassMethod|Function_|Closure $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        if (! $node instanceof Closure) {
+            return $this->processParams($node);
+        }
+
+        $node = $this->processUses($node);
+        return $this->processParams($node);
```
Comment (simivar) [inline rules/downgrade-php80/src/Rector/ClassMethod/DowngradeTrailingCommasInParamUseRector.php:83]:
```suggestion
        if ($node instanceof Closure) {
            $node = $this->processUses($node);
        }
        
        return $this->processParams($node);
```

File: rules/downgrade-php80/src/Rector/ClassMethod/DowngradeTrailingCommasInParamUseRector.php (hunk 1 above)
Comment (simivar) [reply rules/downgrade-php80/src/Rector/ClassMethod/DowngradeTrailingCommasInParamUseRector.php:78]: Done in a separate discussion.

File: rules/downgrade-php80/src/Rector/ClassMethod/DowngradeTrailingCommasInParamUseRector.php (hunk 3 above)
Comment (samsonasik) [reply rules/downgrade-php80/src/Rector/ClassMethod/DowngradeTrailingCommasInParamUseRector.php:83]: Closure can have params, that will make closure with params not updated

Comment (samsonasik) [reply rules/downgrade-php80/src/Rector/ClassMethod/DowngradeTrailingCommasInParamUseRector.php:83]: oh, ok, I understand

File: rules/downgrade-php80/src/Rector/ClassMethod/DowngradeTrailingCommasInParamUseRector.php (hunk 2 above)
Comment (samsonasik) [reply rules/downgrade-php80/src/Rector/ClassMethod/DowngradeTrailingCommasInParamUseRector.php]: implemented.

**Comment by samsonasik** [comment]
@simivar all incorporated 👍

**Comment by samsonasik** [comment]
I moved cleaning trailing comma process to separate method.

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba I think it is ready.

**Comment by TomasVotruba** [comment]
Thank you :+1:

**Comment by simivar** [comment]
Very nice work, thank you @samsonasik 👍

**Pull request merged**